### ISSUE TITLE
Handling Maps

### ISSUE THREAD

**Issue by ratulm** [mbknor/mbknor-jackson-jsonSchema#23]
I have the following simple classes. The top-level class has a Map of another class.

```java
     public class Shape {
        private Map<String, Color> _colors;
  
        public Map<String, Color> getColors() {
           return _colors;
        }
     }
   
     public class Color {
       public String colorName;
    }
```

Information about the internal class completely disappears with mbknor-jackson-jsonSchema:
```json
    {
      "$schema" : "http://json-schema.org/draft-04/schema#",
      "title" : "Shape",
      "type" : "object",
      "additionalProperties" : false,
      "properties" : {
        "colors" : {
          "type" : "object",
          "additionalProperties" : true
        }
      }
    }
```

The jackson-module-jsonScheme produces this instead:
```json
    {
      "type" : "object",
      "id" : "urn:jsonschema:org:batfish:client:Client:Shape",
      "properties" : {
        "colors" : {
          "type" : "object",
          "additionalProperties" : {
            "type" : "object",
            "id" : "urn:jsonschema:org:batfish:client:Client:Color",
            "properties" : {
              "colorName" : {
                "type" : "string"
              }
            }
          }
        }
      }
    }
```

Is there an easy fix to this issue?

PS: Lists are being correctly handled.

**Comment by mbknor** [comment]
I'm not sure how a schema describing something as general as a Map to look like.

Can you suggest what you expect the generated schema to look like?

**Comment by ratulm** [comment]
Is the answer clear to you if the key in the map is a string?

(I am looking at what folks recommend; the answer seems to be some encoding using additionalProperties or patternProperties.)

**Comment by mbknor** [comment]
Since I do not have any specific usecase for this, I think you have to propose something concrete and see how it complies with the Draft 4 spec.

**Comment by ratulm** [comment]
For the class structure above, the following will be ideal (for me):

     {
       "$schema" : "http://json-schema.org/draft-04/schema#",
       "title" : "Shape",
       "type" : "object",
       "additionalProperties" : false,
       "properties" : {
         "colors" : {
           "description" : "hola hola",
           "additionalProperties" : {
             "type" : "object",
             "$ref" : "#/definitions/Color"        
           }
         }
       },
       "definitions" : {
         "Color" : {
           "type" : "object",
           "additionalProperties" : false,
           "properties" : {
             "colorName" : {
               "type" : "string"
             }
           }
         }
       }
     }

According to http://jsonschemalint.com/#/version/draft-04/markup/json, the schema above is valid and the following json is conformant to it. 

     {
       "colors" : {
          "a" : {
             "colorName" : "red"
          },
          "b" : {
             "colorName" : "blue"
          }
       }
     }

Plus, docson does the right thing with the schema; shows the property "colors" as where the value is Color. (That the keys are strings is implied as json only supports strings.)

**Comment by mbknor** [comment]
Looks like you're correct.

From http://json-schema.org/latest/json-schema-validation.html#rfc.section.5.18

> If "additionalProperties" is an object, validate the value as a schema to all of the properties that weren't validated by "properties" nor "patternProperties".

I'll give it a try. Thanks for the research

**Comment by ratulm** [comment]
Perfect, thanks. Happy to help test on my (much more complex) class structures when you have something that I can test drive.

**Comment by mbknor** [comment]
I've implemented improved support for maps which I think results in the jsonSchema you describe.
Please test it out.

**Comment by ratulm** [comment]
Thanks! 

This is working, except for one issue. The schema that is produced has an extra "type" : "object" for "colors":

       [....]
       "properties" : {
         "colors" : {
           "type" : "object",            <<<< this one
           "additionalProperties" : {
       [....]

I am not sure if that is necessary, but it is confusing tools that I use to visualize the schema (docson). Would it be OK remove that? Thanks.

**Comment by ratulm** [comment]
@mbknor Any thoughts on this front? Perhaps the note got buried because the issue was closed. (And thanks for replies on the other issues.)

**Comment by mbknor** [comment]
I'll give it another look

**Comment by ratulm** [comment]
Thanks! In case it helps, my pull request had the minor edits needed for this change.

**Comment by mbknor** [comment]
I do not have a proper fix for this yet, but you might take a look at the new annotation @JsonSchemaInject which is included in the new 1.0.16-release. Maybe it can work as a workaround for you.

**Comment by jekkel** [comment]
I can image #50 might solve this issue. @ratulm Can you check?

**Comment by mbknor** [comment]
I assume this issue has been fixed by #50, at least for now..

**Comment by richturner** [comment]
> Thanks!
> 
> This is working, except for one issue. The schema that is produced has an extra "type" : "object" for "colors":
> 
> ```
>    [....]
>    "properties" : {
>      "colors" : {
>        "type" : "object",            <<<< this one
>        "additionalProperties" : {
>    [....]
> ```
> 
> I am not sure if that is necessary, but it is confusing tools that I use to visualize the schema (docson). Would it be OK remove that? Thanks.

Not Sure this is correct. If type is missing then no type should be inferred and instead the property could be any type. Admittedly the presence of `additionalProperties` infers a type of `object` but I think for correctness it should be included.
